### ISSUE TITLE
Should ImageDecoder::read_image take a &mut [u8] instead of returning a Vec<u8>

### ISSUE THREAD

**Issue by fintelia** [image-rs/image#1029]
I can see a couple of tradeoffs arguing both ways.

**Pros:**
* Consistent interface with ImageDecoderExt::read_rect
* Avoids alignment concerns (EDIT: *for decoder implementations*)
* Sidesteps memory limit considerations by making users do memory allocation

**Cons:**
* Breaking change
* More inconvenient for users since they have to allocate a buffer themselves

**Comment by HeroicKatora** [comment]
> Avoids alignment concerns.

Not sure why it would avoid them? The final buffer, containing the actual subpixel samples, must be aligned correctly for that type. If we want to avoid an extra copy and let the decode write directly to the the final buffer then the target buffer needs to be correctly aligned, regardless of whether it is supplied as `&mut [u8]` or not.

I'm going to add some notes about potentially leveraging the `zerocopy` trait from another api. The user allocating the buffer would be a larger improvement if the decoder could choose to write the intended sample format *directly* instead of requiring bytes in the form of `u8`. Note that the caller can allocate memory correctly for the subpixel type as the sample layout is indicated by the decoder in `color_type` previously.

Basically, `canvas` is intended to offer an api similar to this.

**Comment by fintelia** [comment]
The zerocopy crate has a [`AsBytes::as_bytes_mut()`](https://docs.rs/zerocopy/0.2.8/zerocopy/trait.AsBytes.html#method.as_bytes_mut) method which is implemented for `[u16]`, `[f32]` and so forth. With that, it should be relatively easy for users to interact with the API without needing to do copies by checking the color type and then allocating a buffer accordingly.

**Comment by HeroicKatora** [comment]
`as_bytes_mut()` does the transform `&mut [u16] -> &mut [u8]`. If the decoder interfaces takes that resulting byte slice, it does not solve the issue that the decoder needs to do the inverse transform (which is fallible) whent it wants to write *into* a `&mut [u16]`. And it makes the interface somewhat brittle since the type does not make it clear that the slice should be particularly aligned.

**Comment by HeroicKatora** [comment]
A potential other solution would be a wrapper around a mutable byte slice that associates color types with particular alignment requirements:

```
struct PixelOutput<'a> { into: &'a mut [u8] }

impl<'a> PixelOutput<'a> {
    // Pixel subpixels are always correctly aligned for their own color type.
    pub fn new<P: Pixel>(samples: &'a mut [P::Subpixel]) -> Self { }
    // Fallible construction, no static guarantees
    pub fn with_color(bytes: &'a mut [u8], color: ColorType) -> Option<Self> { }

    // Interface within the decoder: Can assert the output slice is as needed for 
    // the previously indicated color type.
    pub fn as_u16(self) -> Option<&'a mut [u16]> { }
    …
}

impl ImageDecoder {
    pub fn read_into(&mut self, output: PixelOutput) -> ImageResult<()> { }
}
```

**Comment by fintelia** [comment]
> it does not solve the issue that the decoder needs to do the inverse transform (which is fallible) when it produces u16 values due to its color type and wants to write into a &mut [u16]. And it makes the interface somewhat brittle since the type does not make it clear that the slice should be particularly aligned.

I was just imagining that the decoders would use `byteorder` to write the u16's or whatever into the output buffer. That way they could guarantee host endianness and not have to care whether the buffer was aligned. Decoders basically already have to do this to provide an implementation of the `Read` trait which requires them to be able to write into potentially misaligned buffers of u8's.

**Comment by icefoxen** [comment]
In my experience... technically, taking a `&mut [u8]` is better since it lets you reuse your `Vec` if you really need to, but in reality 99% of what someone is going to do is shove the image into a `Vec<u8>` and give it to a GPU to load into a buffer anyway.  The cost of allocating the `Vec` vs. reusing it is probably negligible.
